ddtrace/contrib/asgi/middleware.py

```diff
@@ -111,104 +118,137 @@ async def __call__(self, scope, receive, send):
+                    status_code = message["status"]
+                else:
+                    status_code = None
 
-        tags = _extract_versions_from_scope(scope, self.integration_config)
-        span.set_tags(tags)
+                try:
+                    response_headers = _extract_headers(message)
+                except Exception:
+                    log.warning("failed to extract response headers", exc_info=True)
+                    response_headers = None
 
-        async def wrapped_send(message):
-            if span and message.get("type") == "http.response.start" and "status" in message:
-                status_code = message["status"]
-            else:
-                status_code = None
+                trace_utils.set_http_meta(
+                    span, self.integration_config, status_code=status_code, response_headers=response_headers
+                )
 
-            try:
-                response_headers = _extract_headers(message)
-            except Exception:
-                log.warning("failed to extract response headers", exc_info=True)
-                response_headers = None
+                try:
+                    return await send(message)
+                finally:
+                    # Per asgi spec, "more_body" is used if there is still data to send
+                    # Close the span if "http.response.body" has no more data left to send in the
+                    # response.
+                    if message.get("type") == "http.response.body" and not message.get("more_body", False):
+                        span.finish()
+
+            async def wrapped_blocked_send(message):
+                accept_header = headers.get("accept")
+                if span and message.get("type") == "http.response.start":
+                    message["status"] = 403
+                    message["headers"] = [
+                        (
+                            b"content-type",
+                            b"text/html"
+                            if accept_header and "text/html" in accept_header.lower()
+                            else b"application/json",
+                        ),
+                    ]
+
+                if message.get("type") == "http.response.body":
+                    message["body"] = bytes(appsec_utils._get_blocked_template(accept_header), "utf-8")
+                    message["more_body"] = False
 
-            trace_utils.set_http_meta(
-                span, self.integration_config, status_code=status_code, response_headers=response_headers
-            )
+                try:
+                    return await send(message)
+                finally:
+                    if message.get("type") == "http.response.body":
+                        span.finish()
 
             try:
-                return await send(message)
+                if _request_blocked(span):
+                    return await self.app(scope, receive, wrapped_blocked_send)
```

Comment:
do we want to call into the app still or just start the response right away?

since this way we'll still execute the web request up until the response is starting to be returned, so a malicious action could still take place?